### PR TITLE
feat(backend): add SGLang backend implementation

### PR DESCRIPTION
**Pull request by FionaY728**
## Description
This PR introduces a new backend implementation `SglangBackend` located at `sllm/serve/backends/sglang_backend.py`.

Key highlights:
- Adds integration with `sglang.engine` to support SGLang-based inference
- Introduces a custom patch under `sllm_store/sglang_patch` to accelerate model loading via the SLLM store
- Includes a simple test script (`scripts/simple_sglang_test.py`) and a backend unit test (`tests/backend_test/sglang_backend_test.py`)

## Motivation
This is the initial integration work for supporting [SGLang](https://github.com/InternLM/sglang) in ServerlessLLM.
It enables a lightweight, Python-native backend implementation that communicates directly with the SGLang engine.

To run the pytest please use this command:pytest -v tests/backend_test/sglang_backend_test.py

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## Checklist
- [x] I have read the [CONTRIBUTION](https://github.com/ServerlessLLM/ServerlessLLM/blob/main/CONTRIBUTING.md) guide.
- [x] I have updated the tests (if applicable).
- [ ] I have updated the documentation (if applicable).

### CODE REVIEW

**Comment by JustinTong0323** [comment]
Based on my understanding, we should directly use sglang.engine, instead of the frontend.

**Comment by FionaY728** [comment]
> Based on my understanding, we should directly use sglang.engine, instead of the frontend.

Hi Justin, as mentioned above, I've integrated the sglang.engine.
Please feel free to take a look at this draft PR when you get a chance. Thanks!

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

**Comment by JustinTong0323** [comment]
By the way, there are some unrelated changes in this PR. It is recommended to split them into smaller PRs.

File: sllm/serve/backends/sglang_backend.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+               
+            logger.info("SGLang engine shutdown complete")
+           
+        except Exception as e:
+            logger.error(f"Error during shutdown: {e}")
+        finally:
+            self.status = BackendStatus.UNINITIALIZED
+
+
+    async def chat_completion(self, request_data: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (FionaY728) [reply sllm/serve/backends/sglang_backend.py]: Hi Justin, I have removed the unnecessary functions.

File: sllm/serve/backends/__init__.py [hunk 2]
```diff
@@ -1,3 +1,26 @@
+# # ---------------------------------------------------------------------------- #
```
Comment (JustinTong0323) [inline sllm/serve/backends/__init__.py:1]: duplicate?

File: sllm/serve/backends/useless.py [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (JustinTong0323) [inline sllm/serve/backends/useless.py:1]: clean up?

File: sllm_store/examples/save_vllm_model.py [hunk 4]
```diff
@@ -40,11 +40,17 @@ def _run_writer(input_dir, model_name):
...
             model_path = os.path.join(storage_path, model_name)
-            model_executer = llm_writer.llm_engine.model_executor
+            model_executer = llm_writer.llm_engine.engine_core.model_executor
+
+
             # save the models in the ServerlessLLM format
-            model_executer.save_serverless_llm_state(
-                path=model_path, pattern=pattern, max_size=max_size
-            )
+            llm_writer.llm_engine.save_serverless_llm_state(
```
Comment (JustinTong0323) [inline sllm_store/examples/save_vllm_model.py:47]: use formatter

File: sllm_store/sllm_store/proto/storage_pb2.py [hunk 5]
```diff
@@ -1,68 +0,0 @@
-# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (JustinTong0323) [inline sllm_store/sllm_store/proto/storage_pb2.py:1]: why we change it?

File: sllm_store/sllm_store/server.py [hunk 6]
```diff
@@ -11,7 +11,11 @@
 import os
 
 ctypes.CDLL(os.path.join(sllm_store.__path__[0], "libglog.so"))
-
+try:
+    ctypes.CDLL(os.path.join(sllm_store.__path__[0], "libglog.so"))
+except OSError:
+    # 没装或者没编译 glog，就跳过，不影响后续 gRPC 服务
```
Comment (JustinTong0323) [inline sllm_store/sllm_store/server.py:17]: use English

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

**Comment by JustinTong0323** [comment]
formatting CI not passed

**Comment by FionaY728** [comment]
> formatting CI not passed

I have run with pre-commit run -a, so hopefully no formatting errors

File: sglang_example_config.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (future-xy) [inline sglang_example_config.json:1]: @FionaY728 SGlang example config is under the root dir, pls clean up it.

**Review by future-xy — CHANGES_REQUESTED** [review]

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

**Comment by FionaY728** [comment]
> @FionaY728 Thanks for the PR! Could you please add documentation or at least some test instructions to help reviewers and users run it?

Hi Yao,


Here’s how to test it:

Apply the patch to your sglang codebase as described in the instructions.

Save a model in SLLM format using the provided script (e.g., save_sglang_model.py).

Load （open sllm-store）and validate the SLLM-format model with the following command:

bash
PYTHONPATH=$PWD/sglang/python python sllm_store/examples/load_sglang.py --model-name facebook/opt-1.3b --storage-path ./models
And you can also use the sllm format model to verify the backend pytest:def model_path():
    return "your path"

**Comment by future-xy** [comment]
Thanks!

1. SLLM Store now uses a CLI instead of ad hoc scripts for downloading and converting model checkpoints. Could you integrate your script (e.g., `save_sglang_model.py`) into the SLLM Store CLI?

2. The bash command you shared seems specific to your local setup and may not run reliably in other environments. Please take a look at our documentation and previous PRs for examples of how to provide clean and reproducible instructions.

Let us know if you have any questions!


> > @FionaY728 Thanks for the PR! Could you please add documentation or at least some test instructions to help reviewers and users run it?
> 
> Hi Yao,
> 
> Here’s how to test it:
> 
> Apply the patch to your sglang codebase as described in the instructions.
> 
> Save a model in SLLM format using the provided script (e.g., save_sglang_model.py).
> 
> Load （open sllm-store）and validate the SLLM-format model with the following command:
> 
> bash PYTHONPATH=$PWD/sglang/python python sllm_store/examples/load_sglang.py --model-name facebook/opt-1.3b --storage-path ./models And you can also use the sllm format model to verify the backend pytest:def model_path(): return "your path"

**Comment by FionaY728** [comment]
> Thanks!
> 
> 1. SLLM Store now uses a CLI instead of ad hoc scripts for downloading and converting model checkpoints. Could you integrate your script (e.g., `save_sglang_model.py`) into the SLLM Store CLI?
> 2. The bash command you shared seems specific to your local setup and may not run reliably in other environments. Please take a look at our documentation and previous PRs for examples of how to provide clean and reproducible instructions.
> 
> Let us know if you have any questions!
> 
> > > @FionaY728 Thanks for the PR! Could you please add documentation or at least some test instructions to help reviewers and users run it?
> > 
> > 
> > Hi Yao,
> > Here’s how to test it:
> > Apply the patch to your sglang codebase as described in the instructions.
> > Save a model in SLLM format using the provided script (e.g., save_sglang_model.py).
> > Load （open sllm-store）and validate the SLLM-format model with the following command:
> > bash PYTHONPATH=$PWD/sglang/python python sllm_store/examples/load_sglang.py --model-name facebook/opt-1.3b --storage-path ./models And you can also use the sllm format model to verify the backend pytest:def model_path(): return "your path"

Hello Yao:(sllm-store) fiona@jazz2:~/ServerlessLLM/sllm_store$ sllm-store load --help
Usage: sllm-store load [OPTIONS]

  Loads a model from the sllm-store's storage.

  This command is for loading new models from the sllm-store's local storage.

Options:
  --model TEXT                    Model name from HuggingFace model hub
                                  [required]
  --backend TEXT                  Backend (vllm, transformers, sglang)
                                  [required]
  --adapter-name TEXT             Name of the LoRA adapter
  --precision TEXT                Precision of quantized model. Supports int8,
                                  fp4, and nf4 (transformers only)
  --tensor-parallel-size INTEGER  Tensor parallel size
  --storage-path TEXT             Local path where model is saved
  --help                          Show this message and exit.
(sllm-store) fiona@jazz2:~/ServerlessLLM/sllm_store$ sllm-store save --help
Usage: sllm-store save [OPTIONS]

  Saves a model to the sllm-store's storage.

  This command is for adding new models to the sllm-store's local storage.

Options:
  --model TEXT                    Model name from HuggingFace model hub
                                  [required]
  --backend TEXT                  Backend (vllm, transformers, sglang)
                                  [required]
  --adapter-name TEXT             Name of the LoRA adapter
  --tensor-parallel-size INTEGER  Tensor parallel size
  --local-model-path TEXT         Local path to the model snapshot
  --storage-path TEXT             Local path to save the model
  --skip-save                     Skip saving serverless LLM state (for
                                  debugging, sglang only)
  --help                          Show this message and exit. The command you can run:(sllm-store) fiona@jazz2:~/ServerlessLLM/sllm_store$ sllm-store save --model facebook/opt-1.3b --backend sg
lang --storage-path /home/fiona/ServerlessLLM/sllm_store/examples/models
INFO 07-28 14:18:54 cli.py:325] Saving model facebook/opt-1.3b to /home/fiona/ServerlessLLM/sllm_store/examples/models using sglang backend
INFO 07-28 14:18:56 cli.py:149] 🔄 Downloading model facebook/opt-1.3b from HuggingFace...
INFO 07-28 14:18:56 cli.py:157] 🚀 Initializing SGLang Engine with model path: /home/fiona/.cache/huggingface/hub/models--facebook--opt-1.3b/snapshots/3f5c25d0bc631cb57ac65913f76e22c2dfb61d62
Loading pt checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]
Loading pt checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.76s/it]
Loading pt checkpoint shards: 100% Completed | 1/1 [00:01<00:00,  1.76s/it]

Capturing batches (avail_mem=2.06 GB): 100%|█████████████████████████████████| 4/4 [00:01<00:00,  3.10it/s]
INFO 07-28 14:19:06 cli.py:161]  Saving SLLM state to /home/fiona/ServerlessLLM/sllm_store/examples/models/facebook/opt-1.3b ...
Saving tensors: 100% [======================================================================] 100 %
/home/fiona/miniconda/envs/sllm-store/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py:4631: UserWarning: No device id is provided via `init_process_group` or `barrier `. Using the current device set by the user. 
  warnings.warn(  # warn only once
[rank0]:[W728 14:19:09.322129853 ProcessGroupNCCL.cpp:4718] [PG ID 0 PG GUID 0 Rank 0]  using GPU 0 as device used by this process is currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. You can pecify device_id in init_process_group() to force use of a particular device.
INFO 07-28 14:19:10 cli.py:166]  Copying tokenizer/config files ...
INFO 07-28 14:19:13 cli.py:180] ✅ Done!
INFO 07-28 14:19:13 cli.py:396] Model facebook/opt-1.3b saved successfully to /home/fiona/ServerlessLLM/sllm_store/examples/models
(sllm-store) fiona@jazz2:~/ServerlessLLM/sllm_store$ sllm-store load --model facebook/opt-1.3b --backend sglang --storage-path /home/fiona/ServerlessLLM/sllm_store/examples/models
INFO 07-28 14:20:04 cli.py:441] Loading model facebook/opt-1.3b from /home/fiona/ServerlessLLM/sllm_store/examples/models using sglang backend
DEBUG 07-28 14:20:12 torch.py:137] allocate_cuda_memory takes 0.0027723312377929688 seconds
DEBUG 07-28 14:20:12 client.py:72] load_into_gpu: /home/fiona/ServerlessLLM/sllm_store/examples/models/facebook/opt-1.3b, 6c3cb730-1d50-4f56-92dc-16c5e4ca3768
INFO 07-28 14:20:12 client.py:113] Model loaded: /home/fiona/ServerlessLLM/sllm_store/examples/models/facebook/opt-1.3b, 6c3cb730-1d50-4f56-92dc-16c5e4ca3768
INFO 07-28 14:20:12 torch.py:160] restore state_dict takes 0.0007336139678955078 seconds
INFO 07-28 14:20:12 client.py:117] confirm_model_loaded: /home/fiona/ServerlessLLM/sllm_store/examples/models/facebook/opt-1.3b, 6c3cb730-1d50-4f56-92dc-16c5e4ca3768
INFO 07-28 14:20:13 client.py:125] Model loaded
Capturing batches (avail_mem=2.06 GB): 100%|█████████████████████████████████| 4/4 [00:01<00:00,  2.79it/s]
INFO 07-28 14:20:15 cli.py:503] Model loading time: 11.09s
Prompt: 'Hello, my name is', Generated text: " Albert.... Albert Trapedosa.\nWhat if you're Canadian but German?\nWell I have a genius grandmother from Yosemite in California that lived in the same apartment with her parents. I'm not German is lol"
Prompt: 'The president of the United States is', Generated text: " running to continue Pence's legacy of startling stupidity.  If he loses I don't expect a moment's silence for Trump.\nThe problem is normal human decency too. We are a species that always ceases to pull entirely out of our asses when shit is really bad. It's going to be horrific."
Prompt: 'The capital of France is', Generated text: ' "third place"?\nAmerica ain\'t United! Get yer priorities straight! Stannie law!'
Prompt: 'The future of AI is', Generated text: " defence, its current reliance on offensive technologies will lead to increased adversarial alliances, international legal uncertainty and diminishing international trust in AI — all attempts to the grand stage of the military rivalry. Key Chinese defence surprises include the May 2019 developments in flying drones and helicopter carrier capability, and now China's acquisition and enhancing of advanced cybersecurity technologies have led to what has been described as a 'cyber cold war.' AI will repeatedly do war for the benefit of a dominating superpower. Whatever type of warfare – cyber, conventional or nuclear – AI will deliver greater degrees of sophistication, enormous data volumes, greater operational volume, greater forgone component loss and likely lower"
INFO 07-28 14:20:16 cli.py:564] Model facebook/opt-1.3b loaded successfully from /home/fiona/ServerlessLLM/sllm_store/examples/models

**Comment by JustinTong0323** [comment]
Lint check is not passed.

**Review by JustinTong0323 — COMMENTED** [review]
Also could you make an end-to-end tutorial/doc to tell people how to use the sglang backend, starting from downloading sllm? (I could also test locally)
